File: Mar-18/30-03-18/c++/03-Sumits-subarray-and-subsequence.cpp
#include<bits/stdc++.h>

using namespace std;

int main()
{
    int t;
    cin >> t;
    while(t--)
    {
        int n;
        cin >> n;
        int ar[n+1] = {0};
        int dp[n+1] = {0};
        
        for(int i = 1; i <= n; i++)
        {
            cin >> ar[i];
            dp[i] = ar[i];
            
        }
        
        int maxi = INT_MIN, non_conti = 0;
        for(int i = 1; i <= n; i++)
        {
            dp[i] = max(dp[i-1] + ar[i], ar[i]);
            if(dp[i] > maxi)
                maxi = dp[i];
            if(ar[i] > 0)
                non_conti += ar[i];
        }
        
        //All negative numbers
        if(non_conti == 0)
        {
            non_conti = maxi;
        }
        
        cout << maxi << " " << non_conti << endl;
    }
}

File: Mar-18/30-03-18/python/03-Sumits-subarray-and-subsequence.py
import sys

for _ in range( int(input()) ):
    n = int(input())
    ar = [int(x) for x in input().split()]
    dp = []
    dp.append(0)
    for i in range(0, n):
        dp.append(ar[i])
    
    
    
    maxi = -sys.maxsize
    non_conti = 0
    
    for i in range(1, n+1):
        if(ar[i-1] > 0):
            non_conti += ar[i-1]
            
        dp[i] = max( dp[i-1] + ar[i-1], ar[i-1] )
        if(dp[i] > maxi):
            maxi = dp[i]
        
    
    #All negative numbers
    if(non_conti == 0):
        non_conti = maxi
        
    print(maxi,non_conti)
